kernel/src/engine/default/parquet.rs

```diff
@@ -43,6 +110,63 @@ impl<E: TaskExecutor> DefaultParquetHandler<E> {
         self.readahead = readahead;
         self
     }
+
+    // Write `data` to `path`/<uuid>.parquet as parquet using ArrowWriter and return the parquet
+    // metadata.
+    //
+    // Note: after encoding the data as parquet, this issues a PUT followed by a HEAD to storage in
+    // order to obtain metadata about the object just written.
+    async fn write_parquet(
+        &self,
+        path: &url::Url,
+        data: Box<dyn EngineData>,
+    ) -> DeltaResult<ParquetMetadata> {
+        let batch: Box<_> = ArrowEngineData::try_from_engine_data(data)?;
+        let record_batch = batch.record_batch();
+
+        let mut buffer = vec![];
```

Comment:
It sucks that object_store doesn't have any notion of taking a "stream of bytes" and instead wants everything as a single chunk.

We should probably at least file a tracking issue to see if we can improve this somehow.

Reply:
I think a streaming write to an object store would have to take the form of a multipart upload which `object_store` does support with the [`MultipartUpload`](https://docs.rs/object_store/latest/object_store/trait.MultipartUpload.html) trait and perhaps we can just implement that in some async way to stream our data? also ran across [`WriteMultipart`](https://docs.rs/object_store/latest/object_store/struct.WriteMultipart.html) but I'll do some more digging.

I've already got #418 open for doing streaming JSON puts, and i've just gone ahead and expanded that issue to be streaming JSON/parquet since I think it will generally be the same

Reply:
That `WriteMultipart` does indeed seem relevant. Looks like it's up to us to sort out simple PUT vs. MPU tho (similar to how arrow parquet reads are super low-level, I guess). 

Parquet will be trickier because somebody (engine or default engine) has to encode+compress the bytes that will eventually be written. Parquet is designed to be streamed (row groups, and footer comes last), but I don't know how well arrow-parquet supports it for default client? There's e.g. [SerializedFileWriter](https://arrow.apache.org/rust/parquet/file/writer/struct.SerializedFileWriter.html) that uses [Write](https://doc.rust-lang.org/stable/std/io/trait.Write.html), so I guess we'd have to manually tie that in with the object store machinery (`WriteMultipart` doesn't seem to impl `Write`). 

Also: I _hope_ there's some easy way to write out arrow data as parquet? From my quick skim, it looks like we'd be wrangling individual row groups and column chunks.

Reply:
Yep gonna punt on this for now and optimize in a follow-up.

> Also: I hope there's some easy way to write out arrow data as parquet? From my quick skim, it looks like we'd be wrangling individual row groups and column chunks.

we are just using `parquet::arrow::arrow_writer::ArrowWriter` here! (rather straightforward to use to write record batches - not sure if that's what you were asking)

Reply:
ah yea got it! I can create an issue for this since it's seeming different than the full 'stream to multipart' one?

Reply:
Sorry, too much time invested in my aside here. It would be an issue on `object_store` which we _could_ look at, but probably not now, and probably not something to make an issue for in this repo.